src/passes/OptimizeInstructions.cpp

```diff
@@ -1804,6 +1823,43 @@ struct OptimizeInstructions
       // Don't bother when `ifFalse` isn't pure - we would need to reverse the
       // order using a temp local, which would be bad
     }
+    {
+      // TODO: Remove this after landing SCCP pass. See: #4161
+
+      // i32(x) ? i32(x) : 0  ==>  x
+      Expression *x, *y;
+      if (matches(curr, select(any(&x), i32(0), any(&y))) &&
+          areConsecutiveInputsEqualAndFoldable(x, y)) {
+        return curr->ifTrue;
+      }
+      // i32(x) ? 0 : i32(x)  ==>  { x, 0 }
+      if (matches(curr, select(i32(0), any(&x), any(&y))) &&
+          areConsecutiveInputsEqualAndFoldable(x, y)) {
+        return builder.makeBlock({builder.makeDrop(x), curr->ifTrue});
+      }
+
+      // i64(x) == 0 ? 0 : i64(x)  ==>  x
+      // i64(x) != 0 ? i64(x) : 0  ==>  x
+      if ((matches(curr, select(i64(0), any(&x), unary(EqZInt64, any(&y)))) ||
+           matches(
+             curr,
+             select(any(&x), i64(0), binary(NeInt64, any(&y), i64(0))))) &&
```

Comment:
Hmm, we should optimize `int64(x) != 0` into `!!x` since that is shorter. Let's do that first before this PR, so we don't match the less-optimal pattern here (and then we'd need to fix that up later). I opened #4194 now.

Reply:
Sure!